diff --git a/scripts/build/aflgo-build.sh b/scripts/build/aflgo-build.sh
index 935c2f9..23a5627 100755
--- a/scripts/build/aflgo-build.sh
+++ b/scripts/build/aflgo-build.sh
@@ -68,8 +68,8 @@ python3 -m pip install networkx pydot pydotplus
 export CXX=clang++
 export CC=clang
 # build AFLGo
-cd /afl
+cd /home/maze/tools/afl
 make clean all
 pushd llvm_mode; make clean all; popd
 pushd distance_calculator; cmake -G Ninja ./; cmake --build ./; popd
-export AFLGO=/afl
+export AFLGO=/home/maze/tools/afl
