diff --git a/scripts/build/aflgo-build.sh b/scripts/build/aflgo-build.sh
index c5d958d..8725269 100755
--- a/scripts/build/aflgo-build.sh
+++ b/scripts/build/aflgo-build.sh
@@ -68,9 +68,9 @@ python3 -m pip install networkx pydot pydotplus
 export CXX=clang++
 export CC=clang
 # build AFLGo
-git clone https://github.com/aflgo/aflgo.git
-cd aflgo
+# git clone https://github.com/aflgo/aflgo.git
+cd /home/maze/tools/aflgo
 make clean all
 pushd llvm_mode; make clean all; popd
 pushd distance_calculator; cmake -G Ninja ./; cmake --build ./; popd
-export AFLGO=`pwd`/aflgo
+export AFLGO=/home/maze/tools/aflgo
